provider/throttle_utils.py

```diff
@@ -141,3 +152,600 @@ def change_throttle_group(self, image, group_id):
+    command and check the result of fio command output.
+    Example of usage:
+        ...
+        fio = generate_instance(params, vm, 'fio')
+        tt = ThrottleTester("group1",["img1","img2"])
+        tt.set_fio(fio)
+        tt.build_default_option()
+        tt.build_images_option()
+        tt.start()
+
+    """
+    # Default data struct of expected result.
+    raw_expected = {
+        "burst": {"read": 0, "write": 0, "total": 0, "burst_time": 0,
+                  "burst_empty_time": 0},
+        "normal": {"read": 0, "write": 0, "total": 0}}
+    # Default data struct of raw image data.
+    raw_image_data = {"name": "", "fio_option": "", "output": {}}
+
+    def __init__(self, test, params, vm, session, group, images=None):
+        """
+
+        :param test: Context of test.
+        :param params: params of running ENV.
+        :param vm: VM object.
+        :param session: Session object connect to guest.
+        :param group: Throttle group name.
+        :param images: list of relevant images names.
+        """
+
+        self._test = test
+        self._vm = vm
+        self._session = session
+        self._monitor = vm.monitor
+        self._fio = None
+        self._params = params
+        self.group = group
+        # shared fio option without --filename
+        self._fio_option = ""
+        self.images = images.copy() if images else []
+        self._throttle = {
+            "images": {image: copy.deepcopy(ThrottleTester.raw_image_data) for
+                       image in images},
+            "expected": copy.deepcopy(ThrottleTester.raw_expected)}
+        self._margin = 0.3
+
+    @staticmethod
+    def generate_output_by_json(output):
+        """
+        Convert fio command output to dict object.
+
+        :param output: fio command output with option --output-format=json.
+        :return: dict of fio command output.
+        """
+
+        with tempfile.TemporaryFile(mode="w+") as tmp:
+            tmp.write(output)
+            tmp.seek(0)
+            line = tmp.readline()
+            begin_flag = False
+            block_index = 1
+            block = {}
+            data = ""
+            while line:
+                if line == "{\n":
+                    if not begin_flag:
+                        begin_flag = True
+                    else:
+                        # error
+                        break
+                if begin_flag:
+                    data += line
+                else:
+                    # skip
+                    pass
+                if line == "}\n":
+                    if begin_flag:
+                        begin_flag = False
+                    else:
+                        # error
+                        break
+                    block[block_index] = json.loads(data)
+                    data = ""
+                    block_index += 1
+
+                line = tmp.readline()
+
+            if begin_flag:
+                print("error1")
```

Comment:
for the 'error print', do we need a test.fail("error") here instead of only an error message, or raise exception?
I mean when we hit the error, should we continue the test?

Reply:
It will be fixed return empty dict to indicate something wrong, other function will handle the wrong data.